src/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -180,9 +180,16 @@ private static void initProcessingRules() {
         groupLevelHandlers = new HashMap<String, IElementHandler>() {{
             put("input", new IElementHandler() {
                 @Override public void handle(XFormParser p, Element e, Object parent) {
-                    p.parseControl((IFormElement) parent, e, Constants.CONTROL_INPUT,
-                            Arrays.asList("rows") // Prevent warning about unexpected attributes
-                    );
+                    // Attributes that are passed through to additionalAttributes but shouldn't lead to warnings.
+                    // These are consistently used by clients but are expected in additionalAttributes for historical
+                    // reasons.
+                    final List<String> passedThroughInputAtts = Collections.unmodifiableList(Arrays.asList(
+                            "rows",
+                            "query"
+                    ));
+
+                    p.parseControl((IFormElement) parent, e, Constants.CONTROL_INPUT, passedThroughInputAtts,
+                            passedThroughInputAtts);
```

Comment:
Using `passedThroughInputAtts` for both `additionalUsedAtts` and `passedThroughAtts` bothers me a little. Am I too picky? For your consideration:

![screen shot 2018-03-19 at 5 55 12 pm](https://user-images.githubusercontent.com/55099/37629970-0aaf2c10-2b9f-11e8-88b1-dc499e42655e.png)

Reply:
That’s fine.